streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -804,45 +749,42 @@ private void closeTaskDirty(final Task task) {
         } catch (final RuntimeException swallow) {
             log.error("Error suspending dirty task {} ", task.id(), swallow);
         }
-        cleanupTask(task);
+        tasks.removeTaskBeforeClosing(task.id());
         task.closeDirty();
     }
 
     private void completeTaskCloseClean(final Task task) {
-        cleanupTask(task);
+        tasks.removeTaskBeforeClosing(task.id());
         task.closeClean();
     }
 
-    // Note: this MUST be called *before* actually closing the task
-    private void cleanupTask(final Task task) {
-        for (final TopicPartition inputPartition : task.inputPartitions()) {
-            partitionToTask.remove(inputPartition);
-        }
-    }
-
     void shutdown(final boolean clean) {
         final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
 
         final Set<Task> tasksToCloseDirty = new HashSet<>();
+        // TODO: change type to `StreamTask`
+        final Set<Task> activeTasks = new TreeSet<>(Comparator.comparing(Task::id));
```

Comment:
We need to get a deep-copy of all active tasks here, because `tryCloseCleanAllActiveTasks` would remove tasks from the task container but we need them to iterate over them later to close the task producers.

Reply:
Would it be cleaner to just close/remove the task producer inside `tryCloseCleanAllActiveTasks`? Could we just close/remove the task producers before closing clean or do we specifically need to close them first? I don't remember...

Reply:
Me neither. :) I actually wanted to look into this in more detail in a follow up PR and just added the deep-copy to get this PR into a "correct" state to make the tests pass. If you insist, I can try to update this PR for a proper cleanup, but I am not sure what the scope will be...

Reply:
No worries, let's keep the scope small for now. Just wanted to raise the question

Reply:
Very fair question. Glad that we can follow up on it later.